### PR TITLE
implements missing validations

### PR DESCRIPTION
**Pull request by rashidsp**

### CODE REVIEW

File: tests/test_optimizely.py [hunk 1]
```diff
@@ -1173,25 +1173,28 @@ def test_get_variation__invalid_object(self):
...
+    """ Test that is_feature_enabled returns false if the provided user ID is empty or None. """
 
     opt_obj = optimizely.Optimizely(json.dumps(self.config_dict_with_features))
 
     with mock.patch('optimizely.logger.NoOpLogger.log') as mock_logger:
       self.assertFalse(opt_obj.is_feature_enabled('feature_key', None))
+      self.assertFalse(opt_obj.is_feature_enabled('feature_key', ''))
 
-    mock_logger.assert_called_once_with(enums.LogLevels.ERROR, enums.Errors.NONE_USER_ID_PARAMETER)
+    mock_logger.assert_called_with(enums.LogLevels.ERROR, enums.Errors.INVALID_USER_ID_ERROR)
```
Comment (aliabbasrizvi) [inline tests/test_optimizely.py]: This check only asserts the last call and not both calls. What you can do is `assertFalse`and check log message and then again do the same thing.

Comment (aliabbasrizvi) [reply tests/test_optimizely.py]: I think you have taken a similar approach below.

File: tests/test_optimizely.py [hunk 2]
```diff
@@ -1839,6 +1846,16 @@ def test_activate__dispatch_raises_exception(self):
...
 
+  def test_track__returns_none_if_none_or_empty_user_id(self):
+    """ Test that expected log messages are logged during track when user_id is empty or None. """
+
+    with mock.patch('optimizely.logger.SimpleLogger.log') as mock_logging:
+      self.optimizely.track('test_event', '', event_tags=4200)
+      self.optimizely.track('test_event', None, event_tags=4200)
+
+    mock_logging.assert_called_with(enums.LogLevels.ERROR, enums.Errors.INVALID_USER_ID_ERROR)
+    self.assertEqual(2, mock_logging.call_count)
```
Comment (aliabbasrizvi) [inline tests/test_optimizely.py]: Similar comment as above.

File: tests/test_optimizely.py [hunk 3]
```diff
@@ -1883,6 +1900,16 @@ def test_get_variation__invalid_attributes(self):
...
 
+  def test_get_variation__returns_none_if_none_or_empty_user_id(self):
+    """ Test that expected log messages are logged during get variation when user id is empty or None. """
+
+    with mock.patch('optimizely.logger.SimpleLogger.log') as mock_logging:
+      self.optimizely.get_variation('test_experiment', '', attributes={'test_attribute': 'test_value'})
+      self.optimizely.get_variation('test_experiment', None, attributes={'test_attribute': 'test_value'})
+
+    mock_logging.assert_called_with(enums.LogLevels.ERROR, enums.Errors.INVALID_USER_ID_ERROR)
+    self.assertEqual(2, mock_logging.call_count)
```
Comment (aliabbasrizvi) [inline tests/test_optimizely.py]: Similar comment as above.

**Comment by rashidsp** [comment]
Closing this PR as most of the APIs params validation have been addressed in https://github.com/optimizely/python-sdk/pull/119
However we will create separate PR for params validation of _get_feature_variable_for_type method.

**Pull request closed without merging**